corehq/apps/ota/views.py

```diff
@@ -265,11 +265,20 @@ def get_restore_response(domain, couch_user, app_id=None, since=None, version='1
     app = get_app_cached(domain, app_id) if app_id else None
 
     #For ICDS only allow users who have access the specific app
-    if settings.SERVER_ENVIRONMENT in settings.ICDS_ENVS:
-        role = couch_user.get_role(domain)
-        has_access_to_app = role.permissions.view_web_app(app)
-        if not has_access_to_app:
-            return HttpResponse('User not allowed on this app', status=406), None
+    if (
+        settings.SERVER_ENVIRONMENT in settings.ICDS_ENVS
+        and toggles.ROLE_WEBAPPS_PERMISSIONS.enabled(domain)
+    ):
+        has_access_to_app = False
+        try:
```

Comment:
nit: I prefer keeping it clear about what would raise an exception and combining with couple of other points, my suggestion would be
```
try:
    role = couch_user.get_role(domain)
except DomainMembershipError:
    return HttpResponse('User is not a member of this project', status=404), None
else:
    if not (role and role.permissions.view_web_app(app)):
        return HttpResponse('User is not allowed on this app', status=406), None
```

1. other places where we are raising `DomainMembershipError`, we are returning 404, so this would keep things consistent.
2. I think we can keep the message generic to avoid telling too much to the user or too specific of internals but i will defer to others if they have better suggestions on them.

Reply:
Thanks for the suggestions. I understand your viewpoint and I really liked it.

I have some things to add to the points that you have mentioned. 

For point 1

> other places where we are raising DomainMembershipError, we are returning 404, so this would keep things consistent.

I see the request is already returning 404 when the user is not found, @shubham1g5 just want to confirm that mobile will handle 404 with grace.

For point 2.

> I think we can keep the message generic to avoid telling too much to the user or too specific of internals but i will defer to others if they have better suggestions on them.

The thought process behind specific messages was, we should communicate it to the user why the sync is not happening rather than a generic message. And a plus point will be they can help in faster debugging.
But I am open to suggestions on it.

Reply:
@AmitPhulera  @mkangia `406` is used to communicate a specific user message to the user on mobile, with a `404` you won't be able to do so and mobile will just throw a generic network error for these scenarios.

Reply:
+1 on using a 404 if the user is not part of the domain

401 seems like a more appropriate response in the case that the user does not have access to the app.

Reply:
Thanks for the suggestions.
So I will be going with 404 on the DomainMembership error and I am assuming this will be a rare event if I talk about ICDS.
The snippet @mkangia shared seems to look good would be incorporating that.